extensions/Lily/Assets.js

```diff
@@ -0,0 +1,656 @@
+      const soundIndex = this._getSoundIndexByName(soundName, util);
+      if (soundIndex < 0) return "";
+
+      const sound = target.sprite.sounds[soundIndex];
+      switch (attribute) {
+        case "dataURI":
+          return sound.asset.encodeDataURI();
+        case "index":
+          return soundIndex + 1;
+        case "format":
+          return sound.asset.assetType.runtimeFormat;
+        case "header":
+          return sound.asset.assetType.contentType;
+        case "asset ID":
+          return sound.asset.assetId;
+      }
+    }
+
+    getCostumeAtIndex(args, util) {
+      const target = util.target;
+      const index = Math.round(Cast.toNumber(args.INDEX - 1));
+      const costumes = target.sprite.costumes;
+      if (!(0 < index < costumes.length)) return "";
+
+      return costumes[index].name;
+    }
+
+    getSoundAtIndex(args, util) {
+      const target = util.target;
+      const index = Math.round(Cast.toNumber(args.INDEX - 1));
+      const sounds = target.sprite.sounds;
+      if (!(0 < index < sounds.length)) return "";
+
+      return sounds[index].name;
+    }
+
+    openProject(args) {
+      const url = Cast.toString(args.URL);
+      Scratch.fetch(url)
+        .then((r) => r.arrayBuffer())
+        .then((buffer) => vm.loadProject(buffer));
+    }
+
+    getProjectJSON() {
+      return Scratch.vm.toJSON();
+    }
+
+    loadExtension(args) {
+      const url = Cast.toString(args.URL);
+      vm.extensionManager.loadExtensionURL(url);
+    }
+
+    getLoadedExtensions(args) {
+      return JSON.stringify(
+        Array.from(vm.extensionManager._loadedExtensions.keys())
+      );
+    }
+
+    /* Utility Functions */
+
+    _getSoundIndexByName(soundName, util) {
+      const sounds = util.target.sprite.sounds;
+      for (let i = 0; i < sounds.length; i++) {
+        if (sounds[i].name === soundName) {
+          return i;
+        }
+      }
+      return -1;
+    }
+
+    // PenguinMod
+    _typeIsBitmap(type) {
+      return (
+        type === "image/png" ||
+        type === "image/bmp" ||
+        type === "image/jpg" ||
+        type === "image/jpeg" ||
+        type === "image/jfif" ||
+        type === "image/webp" ||
+        type === "image/gif"
+      );
+    }
+
+    _getTargetFromMenu(targetName, util) {
+      let target = Scratch.vm.runtime.getSpriteTargetByName(targetName);
+      if (targetName === "_myself_") target = util.target.sprite.clones[0];
+      return target;
+    }
+
+    _getTargets() {
+      const spriteNames = [];
+      const targets = Scratch.vm.runtime.targets;
+      for (let index = 1; index < targets.length; index++) {
+        const target = targets[index];
+        if (target.isOriginal) {
+          const targetName = target.getName();
+          const targetId = target.id;
+          spriteNames.push({
+            text: targetName,
+            // We, rather unfortunately, have to rely on Target IDs. This is a terrible idea.
+            value: targetId,

```

Comment:
what reason?

pretty sure those are entirely randomly generated each page load

Reply:
They do, that's a very good point. The problem comes from one of the blocks allowing you to rename a sprite. If the sprite changes names, then the block would suddenly not work so long as it used just the sprite's name.

Reply:
It's the same problem as List Tools, except here it's much more dire because one of the blocks from the extension can literally break itself if it used names as opposed to IDs